ONSdigital/census-rm-case-service
# Distribute case from message consuming thread for performance

# Motivation and Context
The case 'distributor' was very slow. For 300,000 cases it would spend at least 50 minutes simply waiting, doing nothing. The locking mechanism was overly complicated, using a Redis list and a pageable resultset, plus the database as a crude queuing mechanism. It was not scalable and the code was hard to support and maintain.

# What has changed
When a case creation request message arrives, we can send a message to the action service immediately, at the same time as we commit the new case to the database. We use multiple message consuming threads and multiple instances of the case service to achieve a lot of scalability and vastly improve performance.

# How to test?
Load a large sample (e.g. 300k) and check the action case table to see that all the cases are flowing through the case service.

# Links
Trello: https://trello.com/c/cHLbOX2I/606-increase-performance-of-case-distribution
